File: Binary Tree/114M. Flatten Binary Tree to Linked List.cpp
/*
114M. Flatten Binary Tree to Linked List
Tree, Depth-first Search

Description:
Given the root of a binary tree, flatten the tree into a "linked list":

The "linked list" should use the same TreeNode class where the right child pointer points to the next node in the list and the left child pointer is always null.
The "linked list" should be in the same order as a pre-order traversal of the binary tree.

Example 1:
         1
     2       5
  3     4       6

         1 - 2 - 3 - 4 - 5 - 6
Input: root = [1,2,5,3,4,null,6]
Output: [1,null,2,null,3,null,4,null,5,null,6]

Example 2:
Input: root = []
Output: []

Example 3:
Input: root = [0]
Output: [0]

Similar Questions:
Flatten a Multilevel Doubly Linked List - Medium
Correct a Binary Tree - Medium
*/

// Method 1: Recursive approach
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    void flatten(TreeNode* root) {
        if(root==NULL){
            return;
        }
        // Flatten the left tree and the right tree sequentially
        flatten(root->left);
        flatten(root->right);
        
        // Save the flat subtress to temps
        TreeNode* tempLeft = root->left;
        TreeNode* tempRight = root->right;
        // Set the left of root to NULL
        root->left = NULL;
        // Connect the flat left tree to the right of root
        root->right = tempLeft;
        
        // Define a temp node and move it the end of the right tree
        TreeNode* cur = root;
        while(cur->right!=NULL){
            cur = cur->right;
        }
        // Connect the flat right tree to the end of right subtree
        cur->right = tempRight;
    }
};


// Method 2: Iterative approach
class Solution {
public:
    void flatten(TreeNode* root) {
        if(root==NULL){
            return;
        }
        stack<TreeNode*> s;
        s.push(root);
        TreeNode* ref = root;
        while(s.size()!=0){
            TreeNode* cur = s.top();
            s.pop();
            if(cur->right!=NULL){s.push(cur->right);}
            if(cur->left!=NULL){
                s.push(cur->left);
                cur->left = NULL;
            }
            
            // Connect the current node to the right of the previous node
            if(cur!=ref){
                ref->right = cur;
                ref = cur;
            }
        }
        return;
    }
};

File: Binary Tree/114M. Flatten Binary Tree to Linked List.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

# Method 1: Iterative approach
class Solution(object):
    def flatten(self, root):
        """
        :type root: TreeNode
        :rtype: None Do not return anything, modify root in-place instead.
        """
        if root is None:
            return None
        s = collections.deque()
        s.append(root)
        ref = root
        while len(s)!=0:
            cur = s[-1]
            s.pop()
            if cur.right is not None:
                s.append(cur.right)
            if cur.left is not None:
                s.append(cur.left)
                cur.left = None
            if ref!=cur:
                ref.right = cur
                ref = cur
        return root
        
        # Another approach:
        if root is None:
            return root
        s = collections.deque()
        s.append((root, 0))
        parent = root
        while len(s):
            cur, v = s.pop()
            if v==0:
                if cur.right:
                    s.append((cur.right, 0))
                if cur.left:
                    s.append((cur.left, 0))
                    cur.left = None
                s.append((cur, 1))
            else:
                if parent!=cur:
                    parent.right = cur
                    parent = cur
                
        return root


# Method 2: Recursive approach
class Solution(object):
    def flatten(self, root):
        """
        :type root: TreeNode
        :rtype: None Do not return anything, modify root in-place instead.
        """
        if root is None:
            return root
        leftTree = self.flatten(root.left)
        rightTree = self.flatten(root.right)
        root.right = leftTree
        root.left = None
        cur = root
        while cur.right:
            cur = cur.right
        cur.right = rightTree
        return root
